pkg/supportbundle/collect.go

```diff
@@ -207,136 +223,390 @@ func getAnalysisFile(analyzeResults []*analyze.AnalyzeResult) (io.Reader, error)
-		span.SetAttributes(attribute.String("type", reflect.TypeOf(collector).String()))
+		span.SetAttributes(attribute.String("type", "Collect"))
 
 		isExcluded, _ := collector.IsExcluded()
 		if isExcluded {
-			opts.ProgressChan <- fmt.Sprintf("[%s] Excluding host collector", collector.Title())
+			msg := fmt.Sprintf("[%s] Excluding host collector", collector.Title())
+			opts.CollectorProgressCallback(opts.ProgressChan, msg)
 			span.SetAttributes(attribute.Bool(constants.EXCLUDED, true))
 			span.End()
 			continue
 		}
 
-		// Run local collector sequentially
-		result, err := collector.Collect(opts.ProgressChan)
+		// Send progress event: starting the collector
+		msg := fmt.Sprintf("[%s] Running host collector...", collector.Title())
+		opts.CollectorProgressCallback(opts.ProgressChan, msg)
+
+		// convert host collectors into a HostCollector spec
+		spec := createHostCollectorsSpec([]*troubleshootv1beta2.HostCollect{collectorSpec})
+		specJSON, err := json.Marshal(spec)
 		if err != nil {
-			span.SetStatus(codes.Error, err.Error())
-			opts.ProgressChan <- fmt.Sprintf("[%s] Error: %v", collector.Title(), err)
+			return nil, err
 		}
+		klog.V(2).Infof("HostCollector spec: %s", specJSON)
+		for _, pod := range pods.Items {
+			eg.Go(func() error {
+				if err := waitForPodRunning(ctx, clientset, &pod); err != nil {
+					return err
+				}
+
+				stdout, _, err := getExecOutputs(ctx, opts.KubernetesRestConfig, clientset, pod, specJSON)
+				if err != nil {
+					// span.SetStatus(codes.Error, err.Error())
+					msg := fmt.Sprintf("[%s] Error: %v", collector.Title(), err)
+					opts.CollectorProgressCallback(opts.ProgressChan, msg)
+					return errors.Wrap(err, "failed to run remote host collector")
+				}
 
-		// Send progress event: completed successfully
-		opts.ProgressChan <- fmt.Sprintf("[%s] Completed host collector", collector.Title())
+				result := map[string]string{}
+				if err := json.Unmarshal(stdout, &result); err != nil {
+					return err
+				}
 
-		// Aggregate the results
-		for k, v := range result {
-			collectedData[k] = v
+				// Send progress event: completed successfully
+				msg = fmt.Sprintf("[%s] Completed host collector", collector.Title())
+				opts.CollectorProgressCallback(opts.ProgressChan, msg)
+
+				// Aggregate the results
+				mu.Lock()
+				for file, data := range result {
+					if nodeLogs[pod.Spec.NodeName] == nil {
+						nodeLogs[pod.Spec.NodeName] = make(map[string][]byte)
+					}
+					nodeLogs[pod.Spec.NodeName][file] = []byte(data)
+				}
+				mu.Unlock()
+				return nil
+			})
 		}
 
+		err = eg.Wait()
+		if err != nil {
+			return nil, err
+		}
 		span.End()
 	}
-	return nil
+
+	klog.V(2).Infof("All remote host collectors completed")
+
+	defer func() {
+		// TODO:
+		// delete the config map
+		// delete the remote pods
+		// check if the daemonset still exists
+		if err := clientset.AppsV1().DaemonSets(ds.Namespace).Delete(ctx, ds.Name, metav1.DeleteOptions{}); err != nil {
```

Comment:
`ds` could be nil here if there is error before `ds` is is created, we have to do a nil check for `ds` here if not the code may panic.

Reply:
I see. I have added ds nil check and ds name check